### DISCUSSION TITLE
Passwords and encryption section of manual

### DISCUSSION THREAD

**Discussion by milsap1234** [qpdf/qpdf#838] [Q&A]
Hi Jay,
Thanks for writing the new manual section on encryption; it helps me a lot. I have one small suggestion.  Near the beginning of the section and also in --help=encryption, it says "Either or both of user-password and owner-password may be empty strings."  I think newcomers will thank you if you if you incorporate there something you've said on forums:  "Either or both of user-password and owner-password may be empty strings, but in that case, the file can be trivially decrypted." Similarly, in the very interesting statement "The security of a 256-bit AES-encrypted PDF file with a strong password is comparable to using a general-purpose encryption tool like gpg...", "**a** strong password" could be misleading.  

Excuse me for harping on this point--new to qpdf, I made the potentially costly mistake of encrypting with "" for owner password before you set me straight [here.](https://github.com/qpdf/qpdf/discussions/530)  I have a related question.  In that thread you discussed the case where "one of user/owner password is $pw and the other password is some other password that is at least as strong."   Do you imply that it would be insecure use $pw for both?

Thanks again.  
Cheers
Mark

**Comment by jberkenbilt** [comment]
This is excellent feedback. Thanks. I will attempt to clarify and may ping you here and ask you to take a look. It won't be immediate, but it will be before whatever release is next.

Using the same password for user and owner is not considered secure. For a tool like qpdf that doesn't enforce security restrictions, or really any open source since it can be easily modified, the weaker of the two passwords prevails because they both do the same thing. For some commercial tools, if the user password and the owner password match, there becomes no way to modify the security settings because the password is tried first as the user password, and if that works, restrictions are enforced. It's also possible that a tool may go the other way around, though I'm not specifically aware of that case. By that, I mean that if the same password is used, it will be treated as the owner password, in which case none of the restrictions will be enforced.

So....if you are trying to password-protect the file so that the encryption can't be broken, but you don't care about the security restrictions (which are trivial to bypass anyway), then it doesn't matter if you use the same password for user and owner, but you need to use a strong password for both. A better practice would be to use a strong password for each and not to use the same password.

I'll have to reread the gpg comment and see whether clarifying that **both** passwords have to be strong is sufficient or whether I need to dial that back. There are certainly differences between using a general-purpose encryption tool. For example, there is a lot of metadata and structural information you can get from an encrypted file without knowing any of its passwords...things like the number of pages, the page numbers, and stuff like that, though anything that actually constitutes content would be encrypted.

Thanks, and always feel free to raise these issues. I appreciate it when people take the time to tell me how the documentation could have helped them better or clarified some misunderstanding. Hopefully this will save someone else time or trouble.

**Comment by milsap1234** [comment]
I'm glad you found my comment useful. Thanks for another detailed and informative reply.  I'm happy to provide a naive view on any changes you make.  

I understand encryption just one feature of qpdf, but it is what I use the most.  My use case is to post "in case of..." documents on line for my kids, who refuse to put up with gpg.

**Reply by jberkenbilt** [reply]
Yes, definitely the big advantage of PDF encryption is that no external tooling is required to work with the files.

**Comment by milsap1234** [comment]
I'd like to add that after some experiments, I see that qpdf gives clear warnings if insecure encryption is attempted, in most cases. I did see one gap in the defenses:
qpdf --encrypt $pw "" 128 --use-aes=y -- in.pdf out.pdf
succeeds with no warning. But, as the manual says, "there is no reason to use 128-bit encryption with AES, just use 256-bit encryption instead." :)

Thanks!

**Reply by jberkenbilt** [reply]
With aes=y, it's not using an insecure encryption protocol, so that's intentional. The warning are specifically around use of RC4. Certainly 256-bit is better than 128-bit with AES, but 128 with AES is not considered weak in that sense.

**Comment by jberkenbilt** [comment]
For 11.4, I have mentioned about being able to trivially decrypt files without a password if either user or owner password is empty.

**Comment by milsap1234** [comment]
Thanks for keeping this in mind.

--
